core/src/env/api.rs

```diff
@@ -64,8 +68,19 @@ pub unsafe fn load(key: Key) -> Option<Vec<u8>> {
 /// own to always encode the expected type.
 pub unsafe fn r#return<T, E>(value: T) -> !
 where
-    T: parity_codec::Encode,
+    T: Encode,
     E: Env,
 {
     E::r#return(&value.encode()[..])
 }
+
+/// Dispatches a Call into the runtime, for invoking other substrate
+/// modules. Dispatched only after successful contract execution.
+///
+/// The encoded Call MUST be decodable by the target substrate runtime.
```

Comment:
What happens if it isn't? Can we actually crash the runtime with our smart contracts?

Reply:
No just the smart contract execution will fail:

https://github.com/paritytech/substrate/blob/00053422ea9dbdbe7932038c6b58a8c41861ff50/srml/contracts/src/wasm/runtime.rs#L574
https://github.com/paritytech/substrate/blob/5ae532d4f016cfe8874b2e9f76a771b2a5a969dd/srml/contracts/src/exec.rs#L441

Reply:
But please write that down as comment since it is rather important and non trivial to see.

Reply:
Comment added see :point_down:

Reply:
Perfection